---
title: "Motif enrichment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, include = T)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

## Get KHID to KYID mapping
```{r}
# get KH-to-KY table
kh.ky <- read.table('data-raw/KH2012_KY2019.txt',stringsAsFactors = F)
names(kh.ky) <- c('KHID',"KYID")
kh.ky$KHID <- paste0("KH2013:",kh.ky$KHID)
kh.ky$KYID <- paste0("KY2019:",kh.ky$KYID)
```

## Parse genomic features from HT.Gene.gff3
```{r}
library(peakToGene)
htky <- getFeatures('data-raw/HT.Gene.gff3')
```

## Read peak data
```{r}
peaks <- import('data-raw/accessomeKY.bed')

library(RSQLite)
con <- dbConnect(SQLite(),'atacCiona.db')

library(CrobustaSQL)
dbListTables(con)
dbReadTable(con, 'peaksets')
DApeaks <- getPeaksets(con)
dir.apply(DApeaks, "DApeaks", row.names=F, col.names=F, append.date=F)

DApeaks <- sapply(list.files("DApeaks", full.names = T), readLines)
names(DApeaks) <- sub('.*\\/(.*)\\.txt', '\\1', names(DApeaks))

library(dirfns)
dir.apply(lapply(DApeaks, function(x) peaks[peaks$name%in%x]), 'DApeaks', dir.export)
```

## Plot distribution of peak lengths
```{r}
# plot peak lengths
dir.hist(width(peaks),'peaklength_total')
# plot only 99% quantile of peak lengths
dir.hist(width(peaks),'peaklength',quant=.99)
```

## Search peaks for motif matches
By default, getMatches() uses motifs from the CrobustaTFs package. You can also specify your own motifs.
```{r}
# find motifs in each peak
library(BSgenome.Crobusta.HT.KY)
library(CrobustaTFs)
library(tfenrichr)

matches <- getMatches(peaks, Crobusta, 'scores', CrobustaMotifs[1:100], p.cutoff=5e-05)
```

## Find overlaps between peaks and features
```{r}
# find which peaks overlap which features
overlaps <- lapply(htky,getOverlaps,peaks)
# select unique peak IDs in each feature
featpeaks <- sapply(overlaps, function(x) unique(x[,1]))
```

## Regulatory matrices by feature
```{r}
regMat(overlaps$five_prime_UTR, matches)
```

```{r}
tfGene <- lapply(overlaps,regMat,matches)
tfGeneID <- strsplit(ID(CrobustaMotifs[colnames(matches)]),';')
tfGeneID <- sapply(tfGeneID, function(x) sapply(x, function(y) kh.ky[kh.ky[,1]==y,2]))
tfGeneID <- sapply(tfGeneID, paste, collapse=';')
tfGene <- lapply(tfGene, function(x){
	colnames(x) <- tfGeneID
	return(x)
})

dir.apply(tfGene, 'tfToGene', dir.tab, ext='tsv')
dir.apply(tfGene, 'tfToGene', function(x, ...) {
	dir.hist(apply(x,2,sum), ..., xlab='# matches')
})
```

```{r}
peakTF <- as.matrix(motifMatches(matches))*1
colnames(peakTF) <- tfGeneID
dir.tab(peakTF,'peak_TF',ext='tsv')
peakTF
```

```{r}
tmp <- apply(peakTF, 1, function(x) unique(unlist(strsplit(colnames(peakTF)[as.logical(x)], ';'))))
sel <- sapply(tmp,length)>0
tmp <- do.call(rbind, mapply(cbind, PeakID=row.names(peakTF)[sel], KYID=tmp[sel]))
tmp
dir.tab(tmp, 'peakToTF', ext='tsv', row.names=F)
dir.apply(featpeaks, 'peakByFeat',dir.tab, ext='tsv', col.names=F, row.names=F)
```

## Run enrichment test for motifs enriched in genomic features
```{r}
# use poisson test to find motif enrichment in each feature with whole accessome as background
pois <- lapply(featpeaks,matchPois,matches)

#write test results to folder enrichmentByFeature
dir.apply(pois,'enrichmentByFeature')

mat <- lapply(
	c('log2OR', 'padj', 'testMotifPerKbp'),
	function(x) sapply(pois, '[', , x)
)
sel <- apply(mat[[2]] < .10, 1, any)

library(moreComplexHeatmap)

dotPscale(
  mat[[1]][sel,],
  mat[[2]][sel,],
  mat[[3]][sel,],
  outl.name='Motifs per Kbp',
  file='featmotifs',
  append.date=T,
  height=50
)
```

## Run enrichment test for motifs enriched in genomic features
```{r}
# use poisson test to find motif enrichment in each feature with whole accessome as background
DApois <- lapply(DApeaks,matchPois,matches)

#write test results to folder enrichmentByFeature
dir.apply(DApois,'enrichmentByDA')

mat <- lapply(
	c('log2OR', 'padj', 'testMotifPerKbp'),
	function(x) sapply(DApois, '[', , x)
)
sel <- apply(mat[[2]] < .10, 1, any)

hyperDot(
  mat[[1]][sel,],
  mat[[2]][sel,],
  mat[[3]][sel,],
  size.name='Motifs per Kpb',
  file='DATFs',
  append.date=T,
  height=50
)
dotPscale(
  mat[[1]][sel,],
  mat[[2]][sel,],
  mat[[3]][sel,],
  outl.name='Motifs per Kbp',
  file='DATFs2',
  append.date=T,
  height=50
)
dotPscale(
  mat[[1]],
  mat[[2]],
  mat[[3]],
  outl.name='Motifs per Kbp',
  file='DATFs3',
  append.date=T,
  height=50
)
```

## Match 5mers
```{r}
pwms <- getPWMs(5)

kmatches <- getMatches(peaks, Crobusta, 'scores', pwms)
kpois <- lapply(DApeaks,matchPois,matches)
dir.apply(kpois,'enrichmentByDA5mer')

tmp <- apply(
	motifMatches(kmatches), 1, 
	function(x) unique(unlist(strsplit(colnames(kmatches)[as.logical(x)], ';')))
)

sel <- sapply(tmp,length)>0
tmp <- do.call(rbind, mapply(cbind, PeakID=row.names(kmatches)[sel], KYID=tmp[sel]))
dir.tab(tmp, 'peakTo5mer', ext='tsv', row.names=F)

mat <- lapply(
	c('log2OR', 'padj', 'testMotifPerKbp'),
	function(x) sapply(kpois, '[', , x)
)
sel <- apply(mat[[2]] < .10, 1, any)

hyperDot(
  mat[[1]][sel,],
  mat[[2]][sel,],
  mat[[3]][sel,],
  size.name='Motifs per Kpb',
  file='DA5mers',
  append.date=T,
  height=50
)
dotPscale(
  mat[[1]][sel,],
  mat[[2]][sel,],
  mat[[3]][sel,],
  outl.name='Motifs per Kbp',
  file='DA5mers2',
  append.date=T,
  height=50
)
dotPscale(
  mat[[1]],
  mat[[2]],
  mat[[3]],
  outl.name='Motifs per Kbp',
  file='DA5mers3',
  append.date=T,
  height=50
)
```

## Read expression data
```{r}
# read expression data
rnf <- read.csv('data-raw/toptags_rnfko_KH_UniqieName.csv',row.names=2)[,-1]

# merge expression data with KHID
rnfky <- merge(rnf,kh.ky,by.x=0,by.y="KHID")
names(rnfky)[1] <- "KHID"

# select significant genes
sig <- rnfky[rnfky$FDR<.05,]

# merge DE genes with peaks by feature
rnfpeaks <- lapply(overlaps,function(x){
	merge(x,sig,by.x="GeneID",by.y="KYID")
})
```

## Test enrichment for peaks associated to upregulated and downregulated genes
```{r}

upreg <- lapply(rnfpeaks,function(x) x[x$logFC>0,])
downreg <- lapply(rnfpeaks,function(x) x[x$logFC<0,])
bg <- lapply(featpeaks,function(x) matches[x,])

# test for motif enrichment in each feature in DE genes using all peaks in a feature as background
uptest <- mapply(matchPois, lapply(upreg,'[',,"PeakID"), bg, SIMPLIFY=F)
dir.apply(uptest,'upregEnrichment')

downtest <- mapply(matchPois, lapply(downreg,'[',,"PeakID"), bg, SIMPLIFY=F)
dir.apply(downtest,'downregEnrichment')
```

## Plot enrichment results
```{r}
mergeStats <- function(colname, tests){
  sapply(tests,'[',,colname)
}

downmat <- lapply(colnames(downtest[[1]]), mergeStats, downtest)
names(downmat) <- colnames(downtest[[1]])

dotPscale(
  downmat$log2OR,
  downmat$padj,
  downmat$testMotifPerKbp,
  outl.name='Motifs per Kbp',
  file='downregEnrichment',
  append.date=T,
  height=50
)
```

## Plot only significant motifs
```{r}
#limit to enriched motifs (positive OR)
sel <- apply(downmat$log2OR,1, function(x)any(x>0)) & apply(downmat$padj,1, function(x)any(x<.05))
downmat <- lapply(downmat,'[',sel,)

dotPscale(
  downmat$log2OR,
  downmat$padj,
  downmat$testMotifPerKbp,
  outl.name='Motifs per Kbp',
  file='downregEnrichmentSig',
  append.date=T,
  height=15
)

dotPscale(
  downmat$log2OR,
  downmat$padj,
  downmat$testMotifPerKbp/downmat$bgMotifPerKbp,
  outl.name='Motifs per Kbp',
  file='downregEnrichmentScaled',
  append.date=T,
  height=15
)
```

## Rnf peaks by differential accessibilty
```{r}
# merge rnf peaks
alldown <- unique(unlist(lapply(downreg, '[', , 'PeakID')))
lapply(DApeaks, intersect, alldown)
```
